### PR TITLE
T/61: Improved the typing when the whole content is selected

### LINKED ISSUE #61: Backspace on a fully selected content or at the beginning of the first block should ensure that paragraph is left
Like in https://github.com/ckeditor/ckeditor5-engine/issues/670.

The paragraph should be left when a full content was selected or backspace was pressed when selection was at the beginning of the first block.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Improved the typing when the whole content is selected. Closes #61.
* After typing when the entire content was selected – text will be wrapped in the paragraph.
* In an empty editor – press and hold the Backspace key will not remove styles of the current element. The entire content will be removed but the element will left.

### CODE REVIEW

File: src/deleteobserver.js [hunk 1]
```diff
@@ -55,4 +64,6 @@ export default class DeleteObserver extends Observer {
  * @param {module:engine/view/observer/domeventdata~DomEventData} data
  * @param {'forward'|'delete'} data.direction The direction in which the deletion should happen.
  * @param {'character'|'word'} data.unit The "amount" of content that should be deleted.
+ * @param {Number} data.sequence A number that describes which sequence of the same event is fired.
```
Comment (Reinmar) [inline src/deleteobserver.js]: A number describing which subsequent delete event it is without the key being released. If it's 2 or more it means that the key was pressed and hold.

**Comment by Reinmar** [comment]
The PR is really good. I had a bit different image of where the code should land, but it makes a lot of sense what you did (putting the code to the command).

The only thing is... the code duplication with the engine ;| We need to figure out how to reuse the engine because this is quite a lot of code which I don't know where could land otherwise.

So, there are at least two options:

* add an option to `deleteContent()` which will make it skip the "multiple elements selected" check,
* make the code used by `deleteContent()` and this command reusable.

The latter has a lot more sense. But it's tricky because we need to find the right pieces of functionality which make sense in some broader context. Like "is entire content selected" and maybe something more.

File: src/deletecommand.js [hunk 2]
```diff
@@ -92,4 +101,81 @@ export default class DeleteCommand extends Command {
 			this._buffer.unlock();
 		} );
 	}
+
+	/**
+	 * If the user keeps <kbd>Backspace</kbd> or <kbd>Delete</kbd> key, we do nothing because the user can clear
+	 * the whole element without removing them.
+	 *
+	 * But, if the user pressed and released the key, we want to replace the entire content with a paragraph if:
+	 *   - the entire content is selected,
```
Comment (Reinmar) [inline src/deletecommand.js]: Blank line before a list in MD. And start items from `* ` (without additional indentation).

**Comment by Reinmar** [comment]
"get limit element" would be another piece of logic to be exposed somewhere else.

**Comment by Reinmar** [comment]
and finally, in order to leave just a paragraph we can do this:

* select entire content of the root (for real now),
* call `deleteContent()`

In sucj case, `deleteContent()` will really remove all content and leave just a paragraph (if it can be there).

So, these 3 things would allow us removing nearly entire duplication.

**Comment by pomek** [comment]
I introduced some fixes but it still requires some work…

* duplicated checks "_whether the entire content should be replaced with a paragraph_" have been removed,
* API of the private method – `DeleteCommand#_shouldEntireContentBeReplacedWithParagraph()` – was simplified (as you wrote),
* fixes for docs,
* replacing with the paragraph is done by engine,
* an option `skipParentsCheck` in the engine has been added – https://github.com/ckeditor/ckeditor5-engine/compare/t/ckeditor5-typing/61.

Unfortunately, `getLimitElement()` function still occurs twice (in the engine and in the typing).

I am not sure whether I'm following a good way. Maybe should I [extract requires function from `DataController#deleteContent`](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/controller/deletecontent.js#L172-L227) as the utils?

And the one more (and the most important) thing – one of the test does not pass after my changes.

Because the paragraph is not allowed in current element, replacing with the paragraps is not executed and at the end we have an empty document with the selection: [`[]`](https://github.com/ckeditor/ckeditor5-typing/pull/111/commits/ef6a30ebb5f5424aa23a8f27cec2061fafd6f998#diff-26221486fc4fa9bec42de9ae6a13a793R195).

**Comment by pomek** [comment]
I've just created a ticket for your suggestion – https://github.com/ckeditor/ckeditor5-engine/issues/1041. It's required before merging this PR.

**Comment by Reinmar** [comment]
That ticket mentioned 3 methods, out of which only `getLimitElement()` made sense. The other two will not be useful.

From this PR I can see that there are, in general, two methods which might be added to the engine – `getLimitElement()` and `isEntireContentSelected()`. The former will make sense in `Schema`. The latter perhaps in `Selection` (I've been considering also `EditingController` or `Document`).

**Comment by Reinmar** [comment]
Could you refresh this PR so it's ready to review? Now, the build was failing. If there are some issues that you'd like to work on together with me, please describe them.

**Comment by pomek** [comment]
> If there are some issues that you'd like to work on together with me, please describe them.

I guess only the one thing is adding a new method to `Selection` – `Selection#isEntireContentSelected( element )`. Is it correct?

**Comment by Reinmar** [comment]
What do you mean?

**Comment by pomek** [comment]
Can we add a new method to Selection class as you wrote a few days ago? The same as we did it for `getLimitElement()` function (added as a method in `Schema`). I missed the second function.

**Comment by Reinmar** [comment]
Yes, we can.

**Comment by pomek** [comment]
This PR requires https://github.com/ckeditor/ckeditor5-engine/issues/1063.

**Comment by pomek** [comment]
@Reinmar, could you look at this PR once again?

**Comment by Reinmar** [comment]
> I'll fix it.

Done.

But now I noticed that this part is not good:

```js
		// Does nothing if editor already contains an empty paragraph.
		if ( selection.getFirstRange().getCommonAncestor().name === 'paragraph' ) {
			return false;
		}
```

You need to ask yourself a question what exactly you want to check. Because this is not entirely true:

>  other paragraph does not occur in the editor.

**Comment by pomek** [comment]
Fixed. Now it requires https://github.com/ckeditor/ckeditor5-engine/pull/1082.

Btw, something happens with the CI. Builds are not triggered…

**Comment by Reinmar** [comment]
Your last change made the code even more unclear than it should be. You focused on code itself not on behaviour requirements.

What is that common ancestor? Why paragraph can't be inside it? This code is neither correct nor clear. 

Please, before making any changes, think on **WHEN** we shouldn't replace the entire limit element content with a paragraph.

**Comment by pomek** [comment]
TBH I don't understand what I should do.

A part of the code prevents from replacing an empty paragraph with another paragraph. If you want, we can skip it and replace the entire content all the time.

**Comment by Reinmar** [comment]
> A part of the code prevents from replacing an empty paragraph with another paragraph.

First of all – the code says something different:

> * * other empty paragraph does not occur in the limit element.

So, this is the first problem – you're saying now something different than you documented. Which is wrong? The code.

Second – you still didn't describe it precisely. The precise situation is that we don't want to replace this paragraph with a new paragraph:

```html
<limit>
    <p>^</p>
</limit>
```

However, we still want to replace this block quote with a new paragraph:

```html
<limit>
    <blockQuote>
        <p>^</p>
    </blockQuote>
</limit>
```

Will your code do that? I don't think so. Your documentation neither explains what is supposed to happen nor implement it that. Start from writing *precisely* what you want achieve and the code will follow.

**Comment by pomek** [comment]
ATM, the test below even does not work:

```js
it( 'leaves an empty paragraph after removing another paragraph from block element', () => {
	doc.schema.registerItem( 'section', '$block' );
	doc.schema.registerItem( 'blockQuote', '$block' );
	doc.schema.limits.add( 'section' );
	doc.schema.allow( { name: 'section', inside: '$root' } );
	doc.schema.allow( { name: 'blockQuote', inside: 'section' } );
	doc.schema.allow( { name: 'paragraph', inside: 'blockQuote' } );

	setData( doc,
		'<section>' +
			'<blockQuote>' +
				'<paragraph>[]</paragraph>' +
			'</blockQuote>' +
		'</section>'
	);


	editor.execute( 'delete' );

	expect( getData( doc, { selection: true } ) ).to.equal(
		'<section>' +
			'<paragraph>[]</paragraph>' +
		'</section>'
	);
} );
```

because of:

image

And this error occurs even when I remove code which I added.

**Comment by Reinmar** [comment]
Hm... I think I experienced this error in the past.

Yep... https://github.com/ckeditor/ckeditor5-engine/issues/905

Set data without setting selection and then set it using the normal selection API. Or check how I worked around this in block-quote tests.

**Comment by Reinmar** [comment]
BTW, I don't think you need to pass `selection:true` to `getData()`.

**Comment by pomek** [comment]
`setData` works fine. It throws an error when I execute `delete` command. 

image

Something does not work in `mergeBranches` function (in `DeleteController#deleteContent()`).


> BTW, I don't think you need to pass selection:true to getData().

:+1:

**Comment by Reinmar** [comment]
> setData works fine. It throws an error when I execute delete command.

But `DataController#deleteContent()` should not be called at all in this scenario. Should it?

**Comment by pomek** [comment]
Yes, you're right. Thanks for helping! I forgot to allow appending `paragraph` to `section`…

**Comment by pomek** [comment]
Ok, I guess it's ready for review once again.

**Pull request merged**